core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -820,24 +821,64 @@ void finishBecauseUnavailable(ShardId shardId, String message) {
         }
     }
 
+
+    static ConcurrentMap<IndexShardReference, String> openShardReferences;
```

Comment:
I agree that it's more natural there, but I prefer to keep it where it is until there's a clear need for it at the lower level.